messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt

```diff
@@ -0,0 +1,89 @@
+package io.customer.messagingpush.processor
+
+import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import io.customer.messagingpush.MessagingPushModuleConfig
+import io.customer.messagingpush.util.PushTrackingUtil
+import io.customer.sdk.CustomerIOShared
+import io.customer.sdk.data.request.MetricEvent
+import io.customer.sdk.repository.TrackRepository
+
+internal class PushMessageProcessorImpl(
+    private val moduleConfig: MessagingPushModuleConfig,
+    private val trackRepository: TrackRepository
+) : PushMessageProcessor {
+
+    /**
+     * Responsible for storing and updating recent messages in queue atomically.
+     * Once this method is called, the current implementation marks the notification
+     * as processed and future calls for same [deliveryId] will receive true indicating
+     * that the notification is processed already.
+     *
+     * @param deliveryId unique message id received in push payload.
+     * @return true if the message was processed previously; false otherwise.
+     * Callers should generally process notifications only if false was returned.
+     */
+    @Synchronized
```

Comment:
We talked about it before, `ArrayDeque<String>` is not thread-safe, hence we need to add this keyword. We can probably, go for `ConcurrentLinkedDeque` which is thread-safe and then wont even have to use this keyword.  

But any reason, we stuck to `ArrayDeque` ?

Reply:
Also, there is `LinkedBlockingDeque` which could be even better for our use case, thread-safe but allows you to set a maximum size. When you try to add an element while the queue is full, it will automatically remove elements from the other end.

Reply:
I'm okay with even using a simple list that fulfills the purpose here, but `Synchronized` annotations should stay there IMO as we are checking and inserting elements at different lines. And without locking, we may run into race condition resulting in multiple `delivered` metrics.

`LinkedBlockingDeque` looks good to me, I'll take look at other options as well and update this.

Reply:
`Synchronized` and `ArrayDeque` does the same thing as `LinkedBlockingDeque` or `ConcurrentLinkedDeque` without `Synchronized` annotation. 

When multiple threads access a thread-safe collection in Java, the collection will be synchronized, which means that only one thread can access it at a time. This prevents race conditions, which are errors that can occur when multiple threads are trying to access the same data at the same time.

Reply:
Unresolving this conversation because I would like to ask for a conclusion to this conversation. 

In the PR, it looks like `ArrayDequeu` and `@Sychronized` is implemented. I would like to ask the reason for this solution instead of using a different data type that was suggested in this conversation.

Reply:
Me and Rehan discussed this, and the usage of `@Synchronized` is valid in this case, because the critical section (the method in this case) has multiple operations that depend on each other and need to be performed atomically.

Let's dry-run it with multiple threads:

`Thread 1` enters the method and checks if the `deliveryId` is in the `recentMessagesQueue`. It finds that the `deliveryId` is not present. 

`Thread 1` moves on to the else block, and starts adding the `deliveryId` to the `queue`.

At this point, if `Thread 2` is allowed to enter the method, it could also check and find that the deliveryId is not in the recentMessagesQueue (because Thread 1 has not yet finished adding it). This could result in the deliveryId being added to the recentMessagesQueue twice. So, Thread 2 must wait until Thread 1 has completely finished the method.

With the @Synchronized keyword, this problem is avoided because only one thread can execute the method at a time. This is an example of where synchronization is necessary to ensure atomicity of operations.

Regarding the other part, `ArrayDeque` is fine as long as no other method accesses it, which is the case for now. Because the only access point to this array is from a method that is synchronized.

Reply:
Ok. Glad to hear you both talked about it. 

The code as-is in this PR is thread safe, but I do not believe that the code helps prevent miss-use of the `ArrayDequeue` in the future. It would be easy for someone in the future to write code that uses this object and forget to make it thread safe. 

One idea I have to help prevent this is making a thread safe version of `ArrayDequeue`? A class that is a wrapper around a private `ArrayDequeue` object where all access is via `@Synchronized` functions? 

```kt
class ThreadSafeArrayDequeue(maxSize: Int) {

  private dequeue: ArrayDequeue

  @Synchronized 
  get() {}

  @Synchronized
  set() {}
}
```

Sorry the code above is very short. I am writing this before heading out for the weekend. I wanted to try and simply illustrate the idea of a wrapper with `@Synchronized` functions.

Reply:
Thanks Levi. I think using one of the structures suggested above by Shahroz should address your concern as well. I've replaced `ArrayDeque` with `LinkedBlockingDeque` which is thread safe and should work well for us.